### ISSUE TITLE
Being able to read gradients from `register_backward_hook`

### ISSUE THREAD

**Issue by dejanbatanjac** [pytorch/pytorch#24400]
## 🚀 Feature
I assume `register_backward_hook` should catch the gradients. At the moment I was not able to get the gradients, unless reading them from a loop. I am providing the complete example you can run and test.

## Motivation
This would be great feature for telemetry.

## Additional context
I am providing the gist https://gist.github.com/dejanbatanjac/86d02413118cd7fb10b479c1d4c8f318 

cc @ezyang @gchanan @SsnL @albanD

**Comment by zou3519** [comment]
From the documentation:

> The grad_input and grad_output may be tuples if the module has multiple inputs or outputs. The hook should not modify its arguments, but it can optionally return a new gradient with respect to input that will be used in place of grad_input in subsequent computations.

You should be able to print out ginp in your code:

```
def backward_pass(module, ginp, goutp):
    print(ginp)
```

However, your ask is an interesting one. It looks like `module.weight.grad` hasn't been initialized when the backward hook gets called. This may be by design, but it is worth looking into if we want to support this and/or add documentation about it.

**Comment by dejanbatanjac** [comment]
Yes, @zou3519 exactly, tanks for checking this out. It would be nice to set `module.weight.grad` early to the correct non-zero values, so `register_backward_hook` can have the info about the gradients.

Or, as this is exactly the same, the hook callback should be in the spot where gradients are calculated.
If this is not possible, optionally a new hook should be introduced to do so.

This would lead to telemetry elegant solution using hooks. I was able to print out almost anything from the model using the hooks, every layer in and out, any parameter, ... but not the gradients.

**Comment by VitalyFedyunin** [comment]
By @zou3519 : It might be very hard to implement. Downgrading priority.

**Comment by albanD** [comment]
Hi,

Unfortunately we don't plan on supporting this feature for the following reason. If the field is a leaf Tensor, then you can do what you want after the end of the `.backward()` by inspecting your model and the different `.grad` fields (potentially modifying them if you want to). If the field is not a leaf Tensor, then you are doing quite complex stateful nn.Module and you can simply use `register_hook()` on that particular Tensor to get it's gradient (and potentially modify it for the rest of the backward).

In your notebook, you can get the same behavior with the correct gradient values by simply having a function that traverse your model after the `backward()` call the same way you traverse the model to add the hooks.

Feel free to create a new post on the [forum](https://discuss.pytorch.org/) if you have issues getting the info you want from your model !
